src/psyclone/domain/gocean/transformations/kern_trans.py

```diff
@@ -0,0 +1,163 @@
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford STFC Daresbury Lab
+
+'''Raise generic PSyIR representing a kernel-layer routine to
+PSyclone kernel-layer-specific PSyIR which uses specialised classes.
+
+'''
+from psyclone.domain.gocean.kernel import KernelMetadataSymbol
+from psyclone.psyGen import Transformation
+from psyclone.psyir.nodes import Schedule, Container
+from psyclone.psyir.transformations import TransformationError
+
+
+class KernTrans(Transformation):
+    '''Raise a generic PSyIR representation of a kernel-layer routine
+    to a PSyclone version with specialised domain-specific nodes and
+    symbols. This is currently limited to the specialisation of kernel
+    metadata.
+
+    '''
+    def __init__(self):
+        super().__init__()
+        self._metadata_name = None
+
+    def validate(self, node, options=None):
+        '''Validate the supplied PSyIR tree.
+
+        :param node: a PSyIR node that is the root of a PSyIR tree.
+        :type node: Union[:py:class:`psyclone.psyir.node.Routine`, \
+            :py:class:`psyclone.psyir.node.Container`]
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str: str]]
+
+        :raises TransformationError: if the metadata name has not been \
+            set or does not exist in the code.
+        :raises TransformationError: if the supplied node argument is \
+            not a Routine or a Container.
+        :raises TransformationError: if the supplied node argument has \
+            a parent.
+
+        '''
+        if not self._metadata_name:
+            raise TransformationError(
+                "The kern_trans transformation requires the metadata name to "
+                "be set before applying the transformation.")
+
+        metadata_symbol = None
+        for schedule_node in node.walk(Schedule):
+            try:
+                metadata_symbol = schedule_node.symbol_table.lookup(
+                    self._metadata_name)
+                break
+            except KeyError:
+                pass
+        else:
+            raise TransformationError(
+                f"The metadata name ({self._metadata_name}) provided to the "
+                f"transformation does not correspond to a symbol in the "
+                f"supplied PSyIR.")
+
+        # Validate the metadata. This is done as part of the setup()
+        # method. This method can't be called until we specialise the
+        # symbol, so we specialise with a copy of the symbol.
+        tmp_metadata_symbol = metadata_symbol.copy()
+        tmp_metadata_symbol.specialise(KernelMetadataSymbol)
+        tmp_metadata_symbol.setup()
+
+        if not isinstance(node, Container):
```

Comment:
I would have expected these two checks to come first as if they fail, the other tests are unlikely to have passed?